### PR TITLE
[JENKINS-62472] Prevent FINE Empty LogRecorder

### PR DESCRIPTION
**Pull request by Dohbedoh**
See [JENKINS-62472](https://issues.jenkins-ci.org/browse/JENKINS-62472). Prevent users from setting the Root logger to FINE or below from the Log records configuration. This is a very common mistake that user make and cause Jenkins to become very slow, unresponsive and prevent it to start.

### Proposed changelog entries

* [JENKINS-62472] Prevent setting Root Logger to FINE from the Log Recorders configuration

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

-

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I am not sure it should be done in this way. Adding a warning to UI and system logs to **warn** administrators, but at the same time I am not sure why it should be prevented entirely. It is useful when and admin does not know where to look.

Requesting changes to have a discussions

**Comment by res0nance** [comment]
I fully agree with the motivation behind the change, in my organization we have made this mistake on a heavily used master in the past. But then we probably should not prevent this but simply have checks in place to help prevent this from happening by accident instead of intentionally.

**Comment by daniel-beck** [comment]
Could this be done with a warning on form validation when configuring loggers instead?

**Comment by alecharp** [comment]
I agree that we need to do something about this, as it can bring the performance of an instance down easily. However, denying it's configuration entirely doesn't seem the correct fix for me. I'd say having a warning when we are configuring it and a administrative monitor once it's configured would be better (to not forget about it).

**Comment by oleg-nenashev** [comment]
@Dohbedoh Hi. What do you think about the provided feedback? Would you be interested to work on alternate implementation?

**Comment by Dohbedoh** [comment]
@oleg-nenashev Yes I will work on a warning. I was looking for ways to do this with Stapler / Jelly. I am not able to make this work with a `checkUrl` in the `textbox`. I was thinking to add an intermediate screen in that case with a https://stapler.kohsuke.org/apidocs/org/kohsuke/stapler/ForwardToView.html.

**Comment by varyvol** [comment]
@Dohbedoh given there's been no activity for some time now, do you think it makes sense for this to be proposed for close?

**Comment by Dohbedoh** [comment]
@varyvol Will take a look at this today. I could not find simple way to produce an intermediate confirmation page. But found how to fix the problem with the `checkUrl`

**Comment by Dohbedoh** [comment]
Managed to make this work. The textbox needed to be wrapped with a `<f:entry>` for the `checkUrl` warning to work. Here is the result:

enpty-target-name-warning

File: core/src/main/resources/hudson/logging/Messages.properties
Comment (MarkEWaite) [inline core/src/main/resources/hudson/logging/Messages.properties:25]: Optional alternate phrasing
```suggestion
LogRecorder.Target.Empty.Warning=Setting the Root (empty) logger to FINE or below is strongly discouraged. It may critically impact performance. It may crash Jenkins and prevent Jenkins from starting.
```

**Review by MarkEWaite — COMMENTED** [review]
Alternate phrasing proposal.  Optional.

File: core/src/main/resources/hudson/logging/LogRecorder/configure.jelly
Comment (timja) [inline core/src/main/resources/hudson/logging/LogRecorder/configure.jelly:57]: this change needs testing with: https://github.com/jenkinsci/jenkins/pull/3895

nested tables are known to cause issues, it's due for merging into this weekly so let's hold off on this till it's merged

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (timja) [inline core/src/main/java/hudson/logging/LogRecorder.java:159]:
```suggestion
            if((Util.fixEmpty(level) == null || Level.parse(level).intValue() <= Level.FINE.intValue()) 
```

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (timja) [inline core/src/main/java/hudson/logging/LogRecorder.java:165]:
```suggestion
            if (Util.fixEmpty(value) == null) {
```

**Comment by timja** [comment]
on-holding till after tables to divs is merged

**Review by timja — CHANGES_REQUESTED** [review]
It works but the UI is very misaligned for the log level, I also don't think showing the warning message when the user first loads the page is a good user experience

This branch
image

Master:
image

File: core/src/main/resources/hudson/logging/Messages.properties
Comment (daniel-beck) [inline core/src/main/resources/hudson/logging/Messages.properties:25]: There is no natural ordering (lower/higher) to log levels

```suggestion
LogRecorder.Target.Empty.Warning=Setting the Root (empty) logger to a log level more verbose than CONFIG is strongly discouraged. It may critically impact performance. It may crash Jenkins and prevent Jenkins from starting.
```

**Review by daniel-beck — COMMENTED** [review]
A similar warning should be shown in `/log/levels`

File: core/src/main/resources/hudson/logging/LogRecorder/configure.jelly
Comment (Dohbedoh) [reply core/src/main/resources/hudson/logging/LogRecorder/configure.jelly:57]: Will rebase and see how it looks. I guess I should work on converting to `div` as well for that component.

**Comment by alecharp** [comment]
@Dohbedoh The log level misalignment noticed by Tim was not addressed. Do you have time to attend to it? I'm putting the `proposed-for-close` label for the moment, please remove it if you wish to work on the PR.

**Comment by basil** [comment]
I merged and removed the `unresolved-merge-conflict` label. The new test is passing. I think we just need the tables-to-divs update here. @jenkinsci/sig-ux would you be interested in taking this across the finish line? This PR is really close, and I get the sense that the remaining UX work would only take a small amount of time and yield a high benefit for users.

**Review by NotMyFault — COMMENTED** [review]
Thanks for rebasing basil, but I'm unsure if this PR is still needed. The log recorder configuration view already ships a warning about this situation: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/hudson/logging/LogRecorder/configure.jelly#L42

**Comment by basil** [comment]
Thanks!

> I'm unsure

What action is needed for us to be sure?

**Comment by NotMyFault** [comment]
> Thanks!
> 
> > I'm unsure
> 
> What action is needed for us to be sure?

Based on the current layout, I would say this proposal is obsolete, because we have a top level warning in place:

Adding another warning to the input doesn't necessarily trace more attention to it, the top level warning is quite obvious and hard to miss.
I'd vouch for enhancing the top level warning, if needed, by the message of this proposal.

**Comment by basil** [comment]
> Based on the current layout, I would say this proposal is obsolete

I do not agree with the claim that the top-level warning renders this PR obsolete.

The top-level warning addresses a general problem, which is that generating too many FINE logs for a particular package can cause performance and security issues. This warning is of low severity, as there are many legitimate use cases for logging at the FINE level during debugging, and if the FINE logging is constrained to a small set of classes or packages the performance and security issues can be minimized. I would even argue that this would be better as an informational message with a soft color rather than a warning with a yellow/orange/red color because the severity is so low.

In contrast, this PR adds a warning for a _specific_ problem, which is setting the root logger to `FINE` level. Unlike the above use case for setting higher level loggers to FINE, which can be legitimate, setting the root logger to FINE is almost never legitimate and almost always leads to severe performance problems. It is all too easy to shoot oneself in the foot by doing this. We should introduce a guardrail to make it clear that users are about to do something dangerous. This guardrail is of high severity, as failing to heed the warning will almost certainly lead to problems. A warning with a yellow/orange/red color is appropriate here, if the user indeed plans to do something dangerous (probably unintentionally).

In other words, these are two different problems, and the existing top-level warning addresses the first problem, not the second.

>  I'd vouch for enhancing the top level warning, if needed, by the message of this proposal.

I do not agree with the proposal to enhance the top level warning to cover the problem described in this PR. The top level warning is always displayed. This makes sense, because it is of low-severity and applies to a general problem.

In contrast, the warning added in this PR is _specific_ to the root logger, and it is of high severity. Adding a high severity warning should not be taken lightly, and adding such a warning unconditionally is both superfluous and confusing. It would be like warning you to include an @ sign in an email address before you have even typed in the address. If you then proceed to type in the address correctly, you might be annoyed at having been warned not to do something that you were already planning on not doing anyway. In contrast, if you typed in an email address without an @ sign (which would not work) and then were warned about it prior to saving, that would be appropriate, since the system is preventing you from taking a negative action, but only after you indicated an intent to take such a negative action.

The same logic applies here. The problem described in this PR is _specific_ to the root logger, so covering this scenario in a top-level warning that is always displayed is unnecessary and confusing when the root logger has not been set to FINE. Rather, we ought to warn the user not to set the root logger to FINE only in the _specific_ scenario that they have attempted to do that in the first place. And this is what this PR does.

I am requesting that we move forward with this PR, as it introduces a necessary guard rail, and does so only in the case when it is necessary to do so. I continue to request the assistance of the UX SIG to drive this PR through to completion.

**Comment by NotMyFault** [comment]
> In contrast, this PR adds a warning for a specific problem, which is setting the root logger to FINE level. Unlike the above use case for setting higher level loggers to FINE, which can be legitimate, setting the root logger to FINE is almost never legitimate and almost always leads to severe performance problems. It is all too easy to shoot oneself in the foot by doing this. 

Yeah, I agree to disagree with my previous statement :P
This indeed addresses a different concern that should be treated differently.

To follow up on Tim's requested changes, the log level dropdown can be moved to utilize a [modernized component](https://reports.jenkins.io/core-taglib/jelly-taglib-ref.html#form:dropdownDescriptorSelector) instead if a plain selection tag, that way it renders a modernized dropdown component.
I currently have no setup at hand to test this out, but this works out, according to design library.

**Comment by basil** [comment]
Thanks.

> To follow up on Tim's requested changes, the log level dropdown can be moved to utilize a [modernized component](https://reports.jenkins.io/core-taglib/jelly-taglib-ref.html#form:dropdownDescriptorSelector) over a plain selection tag, that way it renders a modernized dropdown component.

Is the lack of a "modernized" component for log level dropdown a blocker for this PR, or is this a pre-existing issue that can be solved independently of this PR (but which this PR would benefit from nonetheless)? If this is a pre-existing issue that can be solved independently, I would like to keep that discussion out of this PR and constrain the scope of this PR to adding the new warning within the existing (not "modernized") log level dropdown. If, on the other hand, the warning introduced in this PR simply cannot be introduced without first "modernizing" the log level dropdown, then I would like to be sure of that.

**Comment by NotMyFault** [comment]
> If this is a pre-existing issue that can be solved independently, I would like to keep that discussion out of this PR and constrain the scope of this PR

Yeah, there are very few occurrences left in core, which use an older dropdown; this is one of them.
This is a pre-existing issue, also visible on the current weekly and can be addressed separately or later on, the warning itself doesn't depend on a modernized component.

**Comment by basil** [comment]
> the warning itself doesn't depend on a modernized component.

Great! This is what I was expecting. Now my remaining request is, within the constraints of the existing (non-modernized) component, can someone from the UX SIG provide a patch that makes this PR feasible for merge (i.e., adding this new warning to the existing component). Please note that this is entirely voluntary, and no pressure if you are unable or unwilling to do this. But I believe it would be low effort, and I believe it is important to drive PRs like this across the finish line to completion and provide a valuable guardrail for Jenkins users. I believe that as core maintainers and UX SIG leaders the responsibility is on us to shepherd the project forward.

**Comment by basil** [comment]
@alecharp Are you interested in taking this PR to the finish line? I think all it needs is to adapt the UI elements to `tables-to-divs`. Seems low-effort and high-reward.

**Comment by timja** [comment]
I have some wip locally to update these controls. Away for the next few days and moving house next week so may be a bit before I can finish it

**Comment by timja** [comment]
I think this PR can be closed in favour of https://github.com/jenkinsci/jenkins/pull/6473

any thoughts?

**Comment by alecharp** [comment]
I think we can close this PR as #6473 is now `ready-for-merge`, right?

**Pull request closed without merging**
